crates/bevy_pbr/src/extended_material.rs

```diff
@@ -97,13 +97,13 @@ pub trait MaterialExtension: Asset + AsBindGroup + Clone + Sized {
 /// When used with `StandardMaterial` as the base, all the standard material fields are
 /// present, so the `pbr_fragment` shader functions can be called from the extension shader (see
 /// the `extended_material` example).
-#[derive(Asset, Clone, TypePath)]
-pub struct ExtendedMaterial<B: Material, E: MaterialExtension> {
+#[derive(Asset, Clone, Reflect)]
+pub struct ExtendedMaterial<B: Material + FromReflect, E: MaterialExtension + FromReflect> {
```

Comment:
Why `FromReflect` bounds here rather than `Reflect`?

Reply:
Not an expert, or even novice on bevy_reflect, but according to the compiler FromReflect seems to be required by TypePath. I tried with Reflect initially, and Reflect + TypePath bounds just now and FromReflect seems to be the thing that works.

Reply:
Interesting. What's the error you're getting if you just add `TypePath` as a bound?

My thought is that we ideally don't want to force reflection on types that don't want or need it (i.e. allow the user to define a `B` and `E` that are not strictly `Reflect`).

Reply:
I think we need reflect on this otherwise we're going to need to do a weird wrapper pattern for any custom material in order to inspect it at runtime. 

StandardMaterial already implements reflect so there's clearly a use for it. 

![image](https://github.com/bevyengine/bevy/assets/77391373/a56d3c72-2b36-47b6-b6ab-5367c1f82c89)

Reply:
What's the code that's causing this error? I ask because I'm wondering if it's an issue with how we're using it rather than how the type is defined. Like, if we're registering concrete types, then there probably shouldn't be an issue. But if we're trying to make the system/plugin generic, then we probably do need these bounds added.

Reply:
Just a 'register_type::<ExtendedMaterial<StandardMaterial,MyCustomMaterial>>>()'

Reply:
I think I found the problem: `Asset` requires `TypePath`, but the `TypePath` impl generated by the `Reflect` derive has additional `FromReflect` bounds, which means that `TypePath` is only implemented where `ExtendedMaterial<B, M>` is `Reflect`.

The fix would be to opt-out of `Reflect`'s automatic `TypePath` implementation and derive it separately:

```rust
#[derive(Asset, Clone, Reflect, TypePath)]
#[reflect(type_path = false)]
pub struct ExtendedMaterial<B: Material, E: MaterialExtension> {
    pub base: B,
    pub extension: E,
}
```

However, this reveals a bug with the `TypePath` derive. Since it reuses some of the `Reflect` derive logic, it actually picks up the `#[reflect(type_path = false)]` attribute and just doesn't generate anything 😅.

As a temporary workaround so this isn't blocked, I recommend doing something like this:

```rust
#[derive(Asset, Clone, Reflect)]
#[reflect(type_path = false)]
pub struct ExtendedMaterial<B: Material, E: MaterialExtension> {
    pub base: B,
    pub extension: E,
}

// We don't use the `TypePath` derive here due to a bug where `#[reflect(type_path = false)]`
// causes the `TypePath` derive to not generate an implementation.
impl_type_path!((in bevy_pbr::extended_material) ExtendedMaterial<B: Material, E: MaterialExtension>);
```

The above code should compile!

Reply:
Oh nice! Would love for this to make it in for .12.1 -- I just pushed the suggested change